edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs

```diff
@@ -97,10 +97,20 @@ static async Task<int> MainAsync(IConfigurationRoot configuration)
             logger.LogInformation("Initializing configuration");
             IConfigSource configSource = await container.Resolve<Task<IConfigSource>>();
             ConfigUpdater configUpdater = await container.Resolve<Task<ConfigUpdater>>();
+            ExperimentalFeatures experimentalFeatures = CreateExperimentalFeatures(configuration);
             var configUpdaterStartupFailed = new TaskCompletionSource<bool>();
-            _ = configUpdater.Init(configSource).ContinueWith(
-                                                        _ => configUpdaterStartupFailed.SetResult(false),
-                                                        TaskContinuationOptions.OnlyOnFaulted);
+            var configDownloadTask = configUpdater.Init(configSource);
+
+            _ = configDownloadTask.ContinueWith(
+                                            _ => configUpdaterStartupFailed.SetResult(false),
+                                            TaskContinuationOptions.OnlyOnFaulted);
+
+            if (!ExperimentalFeatures.IsViaBrokerUpstream(
+                    experimentalFeatures,
+                    string.IsNullOrEmpty(configuration.GetValue<string>(Constants.ConfigKey.GatewayHostname))))
+            {
+                await configDownloadTask;
```

Comment:
I am confused, why can't we await the configUpdater.Init here?

Reply:
The problem is that when edgeHub uses the broker to talk upstream (e.g. to pull the twins), then it needs the protocol heads running.
On the other hand, if protocol heads start up, then clients can start connecting before the configUpdater finishes pulling the config (e.g. the route settings).
This was a bug before (clients started to send messages, but routing was not set), that is why it was waiting before for configUpdater, and then it started the protocol heads.
So now that we want to actieve:
- for 'old' protocol head it should wait to avoid sending messages to early
- for 'new' protocol head, it should spin up before the configUpdater gives up, so it can pull the twin.

The question is that for the 'new' protocol head it isn't a problem that clients connect too early - but it is not, because in this case the client facing interface is the mqtt broker and it will deny client connections until edgehub reports that it is ready. So at the end the system will work for internal communication (e.g. to pull the twin), but clients cannot connect before the twin is pulled for configuration

Reply:
But what about the other protocol heads - AMQP and HTTP?